File: frontend/src/services/interviewService.ts
import api from '@/services/api';
import type {
  Interview,
  InterviewSetup,
  Question,
  Answer,
  InterviewResult,
  AnswerEvaluation,
  SubmitAnswerResponse,
} from '@/types';

const normalizeQuestion = (data: Record<string, unknown>): Question => ({
  id: String(data.id ?? ''),
  text: (data.question_text as string) ?? (data.text as string) ?? '',
  question_text: (data.question_text as string) ?? (data.text as string) ?? '',
  category: (data.question_type as string) ?? (data.category as string) ?? 'technical',
  question_type: (data.question_type as string) ?? (data.category as string) ?? 'technical',
  difficulty: ((data.difficulty as string) ?? 'medium') as 'easy' | 'medium' | 'hard',
  topic: (data.topic as string) ?? undefined,
  expected_focus: (data.expected_focus as string) ?? undefined,
  expectedAnswer: (data.expected_focus as string) ?? undefined,
  parent_question_id: data.parent_question_id ? String(data.parent_question_id) : undefined,
  is_follow_up: Boolean(data.is_follow_up),
  follow_up_reason: (data.follow_up_reason as string) ?? undefined,
});

const normalizeEvaluation = (data: Record<string, unknown>): AnswerEvaluation => ({
  score: (data.score as number) ?? 0,
  relevance_score: (data.relevance_score as number) ?? 0,
  correctness_score: (data.correctness_score as number) ?? 0,
  clarity_score: (data.clarity_score as number) ?? 0,
  depth_score: (data.depth_score as number) ?? 0,
  strengths: Array.isArray(data.strengths) ? (data.strengths as string[]) : [],
  weaknesses: Array.isArray(data.weaknesses) ? (data.weaknesses as string[]) : [],
  missing_points: Array.isArray(data.missing_points) ? (data.missing_points as string[]) : [],
  improvement_feedback: (data.improvement_feedback as string) ?? '',
  ideal_answer_summary: (data.ideal_answer_summary as string) ?? '',
  follow_up_required: Boolean(data.follow_up_required),
  follow_up_reason: (data.follow_up_reason as string) ?? '',
  confidence: (data.confidence as number) ?? 0,
  uncertainty_notes: (data.uncertainty_notes as string) ?? '',
});

const normalizeResult = (data: Record<string, unknown>): InterviewResult => {
  const rawScore = (data.score as number) ?? 0;
  const score = rawScore <= 1 ? Math.round(rawScore * 100) : Math.round(rawScore);
  const normalizeOptionalScore = (key: string) => {
    const v = data[key] as number | null | undefined;
    if (v === null || v === undefined) return null;
    return v <= 1 ? Math.round(v * 100) : Math.round(v);
  };
  return {
    interview_id: data.interview_id ? String(data.interview_id) : undefined,
    score,
    totalQuestions: (data.total_questions as number) ?? undefined,
    answered: (data.answered as number) ?? undefined,
    skipped: (data.skipped as number) ?? undefined,
    feedback: (data.feedback as string) ?? '',
    strengths: Array.isArray(data.strengths) ? (data.strengths as string[]) : [],
    weaknesses: Array.isArray(data.weaknesses) ? (data.weaknesses as string[]) : [],
    suggestions: Array.isArray(data.suggestions) ? (data.suggestions as string[]) : undefined,
    technical_score: normalizeOptionalScore('technical_score'),
    communication_score: normalizeOptionalScore('communication_score'),
    relevance_score: normalizeOptionalScore('relevance_score'),
    problem_solving_score: normalizeOptionalScore('problem_solving_score'),
    resume_alignment: (data.resume_alignment as string) ?? null,
    missing_skills: Array.isArray(data.missing_skills) ? (data.missing_skills as string[]) : undefined,
    preparation_topics: Array.isArray(data.preparation_topics)
      ? (data.preparation_topics as string[])
      : undefined,
    questionResults: Array.isArray(data.question_results)
      ? (data.question_results as Record<string, unknown>[]).map((qr) => ({
          questionId: String(qr.question_id ?? qr.questionId ?? 0),
          question: (qr.question as string) ?? '',
          answer: (qr.answer as string) ?? '',
          score: (qr.score as number) ?? 0,
          feedback: (qr.feedback as string) ?? '',
        }))
      : undefined,
    completion_summary: (data.completion_summary as string) ?? null,
    overall_feedback: (data.overall_feedback as string) ?? null,
    confidence: normalizeOptionalScore('confidence'),
    uncertainty_notes: (data.uncertainty_notes as string) ?? null,
  };
};

const normalizeSetup = (data: Record<string, unknown>, setup?: InterviewSetup): InterviewSetup => {
  if (setup) return setup;
  return {
    resumeId: data.resume_id ? String(data.resume_id) : '',
    jobDescriptionId: data.job_description_id ? String(data.job_description_id) : '',
    difficulty: ((data.difficulty as string) ?? 'medium') as 'easy' | 'medium' | 'hard',
    questionCount: (data.question_count as number) ?? 5,
    type: 'mixed',
    duration: (data.duration as number) ?? 30,
  };
};

const normalizeInterview = (data: Record<string, unknown>, setup?: InterviewSetup): Interview => ({
  id: String(data.id ?? ''),
  user_id: data.user_id ? String(data.user_id) : undefined,
  userId: data.user_id ? String(data.user_id) : undefined,
  setup: normalizeSetup(data, setup),
  title: (data.title as string) ?? 'Mock Interview',
  status: (data.status as string) ?? 'in-progress',
  startedAt: (data.created_at as string) ?? (data.createdAt as string) ?? new Date().toISOString(),
  completedAt: (data.completed_at as string) ?? (data.completedAt as string) ?? undefined,
  questions: Array.isArray(data.questions)
    ? (data.questions as Record<string, unknown>[]).map(normalizeQuestion)
    : [],
  answers: Array.isArray(data.answers) ? (data.answers as Answer[]) : [],
  result: data.result ? normalizeResult(data.result as Record<string, unknown>) : undefined,
});

const questionTypesFromSetup = (setup: InterviewSetup): string[] => {
  if (setup.question_types && setup.question_types.length) return setup.question_types;
  if (setup.type === 'technical') return ['technical'];
  if (setup.type === 'behavioral') return ['behavioral'];
  return ['technical', 'behavioral'];
};

const getToken = () => localStorage.getItem('token') ?? '';

const getByIdInFlight = new Map<string, Promise<Interview>>();
const completeInFlight = new Map<string, Promise<InterviewResult>>();

const cacheKey = (id: string) => `${getToken()}:${id}`;

export const interviewService = {
  create: async (setup: InterviewSetup): Promise<Interview> => {
    const { data } = await api.post('/interviews', {
      title: 'Mock Interview',
      resume_id: setup.resumeId,
      job_description_id: setup.jobDescriptionId,
      difficulty: setup.difficulty,
      question_count: setup.questionCount,
      duration: setup.duration,
      question_types: questionTypesFromSetup(setup),
    });
    return normalizeInterview(data as Record<string, unknown>, setup);
  },

  getById: async (id: string): Promise<Interview> => {
    const key = cacheKey(id);
    const pending = getByIdInFlight.get(key);
    if (pending) return pending;

    const promise = api.get(`/interviews/${id}`).then(({ data }) => {
      getByIdInFlight.delete(key);
      return normalizeInterview(data as Record<string, unknown>);
    });
    getByIdInFlight.set(key, promise);
    return promise;
  },

  getQuestions: async (interviewId: string): Promise<Question[]> => {
    const { data } = await api.get(`/interviews/${interviewId}`);
    return Array.isArray(data.questions)
      ? (data.questions as Record<string, unknown>[]).map(normalizeQuestion)
      : [];
  },

  submitAnswer: async (
    interviewId: string,
    _questionId: string,
    answer: Answer
  ): Promise<SubmitAnswerResponse> => {
    const { data } = await api.post(`/interviews/${interviewId}/answers`, {
      question_id: answer.questionId,
      answer_text: answer.skipped ? '' : answer.text,
      skipped: answer.skipped,
    });
    const raw = (data ?? {}) as Record<string, unknown>;
    return {
      answer: raw.answer as Answer,
      evaluation: raw.evaluation ? normalizeEvaluation(raw.evaluation as Record<string, unknown>) : undefined,
      next_question: raw.next_question
        ? normalizeQuestion(raw.next_question as Record<string, unknown>)
        : undefined,
      follow_up_generated: Boolean(raw.follow_up_generated),
      is_complete: Boolean(raw.is_complete),
      message: (raw.message as string) ?? 'Answer submitted',
    };
  },

  complete: async (interviewId: string): Promise<InterviewResult> => {
    const key = cacheKey(interviewId);
    const pending = completeInFlight.get(key);
    if (pending) return pending;

    const promise = api.post(`/interviews/${interviewId}/complete`).then(({ data }) => {
      completeInFlight.delete(key);
      return normalizeResult(data as Record<string, unknown>);
    });
    completeInFlight.set(key, promise);
    return promise;
  },

  end: async (interviewId: string): Promise<InterviewResult> => {
    return interviewService.complete(interviewId);
  },

  getHistory: async (): Promise<Interview[]> => {
    const { data } = await api.get('/interviews');
    return Array.isArray(data) ? data.map((i: Record<string, unknown>) => normalizeInterview(i)) : [];
  },
};

File: backend/app/services/interview/interview_service.py
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
import uuid

from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import selectinload
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

from app.core.config import get_settings
from app.exception import CustomException
from app.models.interview import Interview as InterviewModel, InterviewStatus
from app.models.interview_answer import InterviewAnswer
from app.models.interview_question import InterviewQuestion, QuestionType
from app.models.interview_result import InterviewResult as InterviewResultModel
from app.schemas.interview import (
    InterviewAnswer as InterviewAnswerSchema,
    InterviewAnswerEvaluation,
    InterviewAnswerResponse,
    InterviewCreate,
    InterviewQuestionResponse,
    InterviewResponse,
    InterviewResult as InterviewResultSchema,
    InterviewResultResponse,
    SubmitAnswerResponse,
)
from app.services.ai.ai_service import AIService
from app.services.interview.interview_query import InterviewQuery
from app.services.job.job_query import JobDescriptionQuery
from app.services.match.match_query import MatchQuery
from app.services.resume.resume_query import ResumeQuery


_MAX_ANSWER_LENGTH = 20000


class Interview:
    def __init__(self, db: AsyncSession):
        self._db = db
        self._query = InterviewQuery(db)
        self._ai = AIService()

    async def _ensure_resume_analysis(self, resume) -> dict:
        if resume.analysis:
            return resume.analysis
        analysis = await self._ai.analyze_resume(resume.extracted_text or "")
        resume.analysis = analysis.model_dump()
        self._db.add(resume)
        await self._db.commit()
        await self._db.refresh(resume)
        return resume.analysis

    async def _ensure_jd_analysis(self, job) -> dict:
        if job.analysis:
            return job.analysis
        analysis = await self._ai.analyze_job_description(job.description or "")
        job.analysis = analysis.model_dump()
        self._db.add(job)
        await self._db.commit()
        await self._db.refresh(job)
        return job.analysis

    async def _load_context(
        self, interview: InterviewModel
    ) -> tuple[dict, dict, Optional[dict]]:
        resume_analysis: dict = {}
        jd_analysis: dict = {}
        match_data: Optional[dict] = None

        if interview.resume_id:
            resume = await ResumeQuery(self._db).get_by_id(interview.resume_id)
            if resume:
                resume_analysis = resume.analysis or {}
        if interview.job_description_id:
            job = await JobDescriptionQuery(self._db).get_by_id(
                interview.job_description_id
            )
            if job:
                jd_analysis = job.analysis or {}
                match = await MatchQuery(self._db).get_by_resume_jd(
                    interview.user_id, interview.resume_id, interview.job_description_id
                )
                if match:
                    match_data = {
                        "overall_match_score": match.overall_match_score,
                        "matched_skills": match.matched_skills,
                        "missing_skills": match.missing_skills,
                        "strengths": match.strengths,
                        "gaps": match.gaps,
                        "recommendations": match.recommendations,
                    }
        return resume_analysis, jd_analysis, match_data

    async def _get_next_question(
        self, interview: InterviewModel
    ) -> Optional[InterviewQuestion]:
        stmt = (
            select(InterviewAnswer)
            .where(InterviewAnswer.interview_id == interview.id)
        )
        result = await self._db.exec(stmt)
        answered_ids = {a.question_id for a in result.all()}

        stmt = (
            select(InterviewQuestion)
            .where(InterviewQuestion.interview_id == interview.id)
            .order_by(InterviewQuestion.order)
        )
        result = await self._db.exec(stmt)
        questions = result.all()
        for q in questions:
            if q.id not in answered_ids:
                return q
        return None

    async def _count_follow_ups(self, interview: InterviewModel) -> int:
        return sum(1 for q in interview.questions if q.is_follow_up)

    @staticmethod
    def _question_to_dict(question: InterviewQuestion) -> Dict[str, Any]:
        return {
            "id": str(question.id),
            "question_text": question.question_text,
            "question_type": question.question_type.value,
            "difficulty": question.difficulty,
            "topic": question.topic,
            "expected_focus": question.expected_focus,
            "order": question.order,
            "category": question.category,
        }

    async def _evaluate_answer(
        self,
        interview: InterviewModel,
        question: InterviewQuestion,
        answer: InterviewAnswer,
    ) -> Optional[InterviewAnswerEvaluation]:
        resume_analysis, jd_analysis, _ = await self._load_context(interview)
        try:
            evaluation = await self._ai.evaluate_answer(
                self._question_to_dict(question),
                answer.answer_text,
                resume_analysis,
                jd_analysis,
                interview.difficulty,
            )
            return InterviewAnswerEvaluation.model_validate(
                evaluation.model_dump(mode="json")
            )
        except CustomException:
            raise
        except Exception:
            return None

    async def _generate_follow_up(
        self,
        interview: InterviewModel,
        question: InterviewQuestion,
        answer: InterviewAnswer,
        evaluation: InterviewAnswerEvaluation,
    ) -> Optional[InterviewQuestion]:
        resume_analysis, jd_analysis, _ = await self._load_context(interview)
        try:
            follow_up = await self._ai.generate_follow_up_question(
                self._question_to_dict(question),
                answer.answer_text,
                evaluation.model_dump(mode="json"),
                resume_analysis,
                jd_analysis,
                interview.difficulty,
            )
        except Exception:
            return None

        max_order = max((q.order for q in interview.questions), default=-1)
        raw_type = (follow_up.question_type or "technical").lower()
        q_type = QuestionType.BEHAVIORAL
        if raw_type == "technical":
            q_type = QuestionType.TECHNICAL

        follow_up_question = InterviewQuestion(
            interview_id=interview.id,
            question_text=follow_up.question,
            question_type=q_type,
            difficulty=follow_up.difficulty or interview.difficulty,
            order=max_order + 1,
            topic=follow_up.topic,
            expected_focus=follow_up.expected_focus,
            category=raw_type,
            parent_question_id=question.id,
            is_follow_up=True,
            follow_up_reason=evaluation.follow_up_reason,
        )
        self._db.add(follow_up_question)
        await self._db.commit()
        await self._db.refresh(follow_up_question)
        return follow_up_question

    async def create_interview(
        self, user_id: uuid.UUID, payload: InterviewCreate
    ) -> InterviewResponse:
        if payload.resume_id:
            resume = await ResumeQuery(self._db).get_by_id(payload.resume_id)
            if not resume or resume.user_id != user_id:
                raise CustomException(404, "Resume not found")
            resume_analysis = await self._ensure_resume_analysis(resume)
        else:
            resume_analysis = {}

        if payload.job_description_id:
            job = await JobDescriptionQuery(self._db).get_by_id(
                payload.job_description_id
            )
            if not job or job.user_id != user_id:
                raise CustomException(404, "Job description not found")
            job_analysis = await self._ensure_jd_analysis(job)
        else:
            job_analysis = {}

        interview = await self._query.create(payload, user_id)
        count = max(1, min(20, payload.question_count))
        questions = await self._ai.generate_interview_questions(
            resume_analysis,
            job_analysis,
            payload.difficulty,
            count,
            payload.question_types or ["technical"],
        )
        for idx, q in enumerate(questions):
            raw_type = (q.get("question_type") or "technical").lower()
            q_type = QuestionType.BEHAVIORAL
            if raw_type == "technical":
                q_type = QuestionType.TECHNICAL
            db_question = InterviewQuestion(
                interview_id=interview.id,
                question_text=q["question"],
                question_type=q_type,
                difficulty=q.get("difficulty", "medium"),
                order=idx,
                topic=q.get("topic"),
                expected_focus=q.get("expected_focus"),
                category=raw_type,
            )
            self._db.add(db_question)
        interview.status = InterviewStatus.IN_PROGRESS
        await self._db.commit()
        await self._db.refresh(interview)
        stmt = (
            select(InterviewQuestion)
            .where(InterviewQuestion.interview_id == interview.id)
            .order_by(InterviewQuestion.order)
        )
        result = await self._db.exec(stmt)
        interview.questions = result.all()
        return InterviewResponse.model_validate(interview)

    async def list_interviews(
        self, user_id: uuid.UUID
    ) -> List[InterviewResponse]:
        interviews = await self._query.get_all(user_id)
        return [
            InterviewResponse.model_validate(i) for i in interviews
        ]

    async def get_interview(
        self, user_id: uuid.UUID, interview_id: uuid.UUID
    ) -> InterviewResponse:
        interview = await self._get_owned_interview(user_id, interview_id)
        return InterviewResponse.model_validate(interview)

    async def submit_answer(
        self,
        user_id: uuid.UUID,
        interview_id: uuid.UUID,
        answer: InterviewAnswerSchema,
    ) -> SubmitAnswerResponse:
        interview = await self._get_owned_interview(user_id, interview_id)
        if interview.status == InterviewStatus.COMPLETED:
            raise CustomException(400, "Interview is already completed")

        question_ids = {q.id for q in interview.questions}
        question_id = answer.question_id
        if question_id not in question_ids:
            raise CustomException(
                400, "Question does not belong to this interview"
            )

        existing = await self._query.get_answer_for_question(
            interview.id, question_id
        )
        if existing:
            raise CustomException(400, "Answer already submitted for this question")

        answer_text = answer.answer_text.strip()
        if not answer_text and not answer.skipped:
            raise CustomException(400, "Answer cannot be empty")
        if len(answer.answer_text) > _MAX_ANSWER_LENGTH:
            raise CustomException(400, "Answer exceeds maximum length")

        question = next(q for q in interview.questions if q.id == question_id)
        db_answer = InterviewAnswer(
            interview_id=interview.id,
            question_id=question_id,
            answer_text=answer_text,
        )
        self._db.add(db_answer)
        await self._db.commit()
        await self._db.refresh(db_answer)

        evaluation_schema: Optional[InterviewAnswerEvaluation] = None
        try:
            evaluation_schema = await self._evaluate_answer(
                interview, question, db_answer
            )
            if evaluation_schema:
                db_answer.score = float(evaluation_schema.score)
                db_answer.evaluation = evaluation_schema.model_dump(mode="json")
                db_answer.evaluated_at = datetime.now(timezone.utc)
                self._db.add(db_answer)
                await self._db.commit()
                await self._db.refresh(db_answer)
        except CustomException:
            pass

        follow_up_generated = False
        if evaluation_schema and evaluation_schema.follow_up_required:
            follow_up_count = await self._count_follow_ups(interview)
            settings = get_settings()
            max_follow_ups = max(0, settings.MAX_FOLLOW_UP_QUESTIONS)
            if follow_up_count < max_follow_ups:
                follow_up = await self._generate_follow_up(
                    interview, question, db_answer, evaluation_schema
                )
                if follow_up:
                    follow_up_generated = True
                    interview = await self._query.get_by_id(interview.id)

        if not follow_up_generated:
            await self._db.refresh(interview)
        next_q = await self._get_next_question(interview)
        is_complete = next_q is None and not follow_up_generated

        return SubmitAnswerResponse(
            answer=InterviewAnswerResponse.model_validate(db_answer),
            evaluation=evaluation_schema,
            next_question=InterviewQuestionResponse.model_validate(next_q)
            if next_q
            else None,
            follow_up_generated=follow_up_generated,
            is_complete=is_complete,
            message="Answer submitted",
        )

    async def complete_interview(
        self, user_id: uuid.UUID, interview_id: uuid.UUID
    ) -> InterviewResultResponse:
        interview = await self._get_owned_interview(user_id, interview_id)
        if interview.result is not None:
            return InterviewResultResponse.model_validate(interview.result)

        report = await self._generate_report(interview)
        result = InterviewResultModel(
            interview_id=interview.id,
            score=report.score,
            feedback=report.feedback,
            strengths=report.strengths,
            weaknesses=report.weaknesses,
            technical_score=report.technical_score,
            communication_score=report.communication_score,
            relevance_score=report.relevance_score,
            problem_solving_score=report.problem_solving_score,
            resume_alignment=report.resume_alignment,
            missing_skills=report.missing_skills,
            suggestions=report.suggestions,
            preparation_topics=report.preparation_topics,
            question_results=report.question_results,
            completion_summary=report.completion_summary,
            overall_feedback=report.overall_feedback,
            confidence=report.confidence,
            uncertainty_notes=report.uncertainty_notes,
        )
        self._db.add(result)
        interview.status = InterviewStatus.COMPLETED
        interview.completed_at = datetime.now(timezone.utc)
        try:
            await self._db.commit()
        except IntegrityError:
            await self._db.rollback()
            existing = await self._query.get_by_id(interview.id)
            if existing is None or existing.result is None:
                raise
            return InterviewResultResponse.model_validate(existing.result)
        await self._db.refresh(result)
        return InterviewResultResponse.model_validate(result)

    async def _generate_report(
        self, interview: InterviewModel
    ) -> InterviewResultSchema:
        resume_analysis, jd_analysis, match_data = await self._load_context(interview)
        qa_data: List[Dict[str, Any]] = []
        for answer in interview.answers:
            question = next(
                (q for q in interview.questions if q.id == answer.question_id), None
            )
            qa_data.append(
                {
                    "question_id": str(answer.question_id),
                    "question": question.question_text if question else "",
                    "answer": answer.answer_text,
                    "evaluation": answer.evaluation or {},
                    "score": answer.score,
                }
            )

        config = {
            "difficulty": interview.difficulty,
            "question_count": interview.question_count,
            "duration": interview.duration,
            "question_types": interview.question_types,
            "answered_count": len(interview.answers),
            "total_questions": len(interview.questions),
        }

        try:
            report = await self._ai.generate_interview_report(
                config,
                resume_analysis,
                jd_analysis,
                match_data or {},
                qa_data,
            )
        except Exception:
            answered = len(interview.answers)
            total = max(1, len(interview.questions))
            avg_score = 0.0
            if answered:
                scores = [a.score for a in interview.answers if a.score is not None]
                avg_score = sum(scores) / len(scores) if scores else 0.0
            answered = len(interview.answers)
            skipped = 0
            total = len(interview.questions)
            return InterviewResultSchema(
                interview_id=interview.id,
                score=avg_score,
                feedback="Final report generation is unavailable. Based on submitted answers, additional review may be needed.",
                strengths=[],
                weaknesses=["Report generation failed"],
                confidence=0.0,
                uncertainty_notes="Could not generate a full report at this time.",
                total_questions=total,
                answered=answered,
                skipped=skipped,
            )

        return InterviewResultSchema(
            interview_id=interview.id,
            score=float(report.overall_score),
            feedback=report.overall_feedback or report.answer_quality_summary,
            strengths=report.strengths,
            weaknesses=report.weaknesses,
            technical_score=float(report.technical_score)
            if report.technical_score is not None
            else None,
            communication_score=float(report.communication_score)
            if report.communication_score is not None
            else None,
            relevance_score=float(report.relevance_score)
            if report.relevance_score is not None
            else None,
            problem_solving_score=float(report.problem_solving_score)
            if report.problem_solving_score is not None
            else None,
            resume_alignment=report.resume_alignment,
            missing_skills=report.missing_or_weak_skills,
            suggestions=report.recommended_preparation_topics,
            preparation_topics=report.recommended_preparation_topics,
            question_results=[q.model_dump(mode="json") for q in report.question_wise_summary],
            completion_summary=report.interview_completion_summary,
            overall_feedback=report.overall_feedback,
            confidence=float(report.confidence),
            uncertainty_notes=report.uncertainty_notes,
            total_questions=len(interview.questions),
            answered=sum(1 for a in interview.answers if a.answer_text.strip()),
            skipped=sum(1 for a in interview.answers if not a.answer_text.strip()),
        )

    async def _get_owned_interview(
        self, user_id: uuid.UUID, interview_id: uuid.UUID
    ) -> InterviewModel:
        interview = await self._query.get_by_id(interview_id)
        if not interview or interview.user_id != user_id:
            raise CustomException(404, "Interview not found")
        return interview
